[6.x] Core/Scripts/BFD: Subjugator Korul - Darkness Calls

tkrokli opened the pull request:
Subjugator Korul spell 151159: Darkness Calls dummy effect,
cast Darkness Calls, summon Deep Terror `<Aspect of Aku'mai>`

Original script written by Ulduar `<Ulduar@skywall.com>` <sub>(account deleted)<sub>

Closes #15316

jackpoz commented on src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:
```diff
@@ -238,10 +239,41 @@ class npc_morridune : public CreatureScript
+
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (Creature* caster = GetCaster()->ToCreature())
```
is there a particular need to check if the caster is a creature ?

tkrokli replied:
Hmm, good question. The original author is not available (user deleted).
If you have a better suggestion, I'm free to change it (unless @Rushor wants it like this).

Shauren replied:
If you need the caster to be a creature (hint: you don't in this simple script) you should enforce it with a typeid check in Load(), not here

Rushor commented on src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:
```diff
@@ -238,10 +239,41 @@ class npc_morridune : public CreatureScript
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (Creature* caster = GetCaster()->ToCreature())
+                {
```
{} not needed here :) 
btw do you know if the spell spell_subjugator_korul_darkness_calls is already defined in spell_script_names in the base?

jackpoz commented on src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:
```diff
@@ -238,10 +239,41 @@ class npc_morridune : public CreatureScript
+            {
+                if (Creature* caster = GetCaster()->ToCreature())
+                {
+                    caster->CastSpell(target, uint32(GetEffectValue()))
```
travis says "target" variable is not declared, which makes one wonder if this script was tested since it doesn't build

tkrokli replied:
OK, I was too triggerhappy when I made this. I thought Rushor, Kilyana, Shauren and Ulduar had worked out how it should be already. Might as well close it for now until I can get it tested.

Rushor replied:
caster->CastSpell(GetHitUnit(), uint32(GetEffectValue()))

try this, better working it out here than closing the pr again, else #15316 will be open for another 2 years

Commit:
[6.x] Core/Scripts/BFD: Subjugator Korul - Darkness Calls

Subjugator Korul spell 151159: Darkness Calls dummy effect,
cast Darkness Calls, summon Deep Terror <Aspect of Aku'mai>

Original script written by Ulduar <Ulduar@skywall.com> (account deleted)

Closes #15316

MitchesD commented on src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:
```diff
@@ -238,10 +239,43 @@ class npc_morridune : public CreatureScript
+                if (Unit* hitUnit = GetHitUnit())
+                    if (Unit* caster = GetCaster())
+                        if (Creature* creature = caster->ToCreature())
+                        {
```
useless brackets

tkrokli commented:
Suggestion: maybe add the labels [Comp-C++Script] + [Comp-Database] to this PR?

Aokromes commented:
Any reason to don't merge this?

Aokromes commented:
If there is no reason to merge this, i will merge it on 24h.

joschiwald commented on src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:
```diff
@@ -238,10 +239,41 @@ class npc_morridune : public CreatureScript
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (Unit* hitUnit = GetHitUnit())
+                    if (Unit* caster = GetCaster())
```
caster is never null in spellscripts

tkrokli replied:
OK, so I just remove line 254, right?

joschiwald commented on src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:
```diff
@@ -238,10 +239,41 @@ class npc_morridune : public CreatureScript
+            {
+                if (Unit* hitUnit = GetHitUnit())
+                    if (Unit* caster = GetCaster())
+                        if (Creature* creature = caster->ToCreature())
```
useless cast

tkrokli replied:
OK, I will remove it.

Aokromes commented:
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Kalimdor/BlackfathomDeeps/blackfathom_deeps.cpp:254:21: fatal error: 
      use of undeclared identifier 'creature'
                    creature->CastSpell(hitUnit, uint32(GetEffectValue()));

Ulduar commented:
Periodic Dummy Templates if need ofc:
``` CPP
class spell_gen_periodic_dummy_aura_caster_on_target : public SpellScriptLoader
{
    public:
        spell_gen_periodic_dummy_aura_caster_on_target() : SpellScriptLoader("spell_gen_periodic_dummy_aura_caster_on_target") { }

        class spell_gen_periodic_dummy_aura_caster_on_target_AuraScript : public AuraScript
        {
            PrepareAuraScript(spell_gen_periodic_dummy_aura_caster_on_target_AuraScript);

            void PeriodicTick(AuraEffect const* /*aurEff*/)
            {
                if (Unit* caster = GetCaster())
                    caster->CastSpell(GetHitUnit(), uint32(GetEffectValue()), true);
            }

            void Register() override
            {
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_caster_on_target_AuraScript::PeriodicTick, EFFECT_0, SPELL_AURA_PERIODIC_DUMMY);
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_caster_on_target_AuraScript::PeriodicTick, EFFECT_1, SPELL_AURA_PERIODIC_DUMMY);
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_caster_on_target_AuraScript::PeriodicTick, EFFECT_2, SPELL_AURA_PERIODIC_DUMMY);
            }
        };

        AuraScript* GetAuraScript() const override
        {
           return new spell_gen_periodic_dummy_aura_caster_on_target_AuraScript();
        }
};

class spell_gen_periodic_dummy_aura_target_on_target : public SpellScriptLoader
{
    public:
        spell_gen_periodic_dummy_aura_target_on_target() : SpellScriptLoader("spell_gen_periodic_dummy_aura_target_on_target") { }

        class spell_gen_periodic_dummy_aura_target_on_target_AuraScript : public AuraScript
        {
            PrepareAuraScript(spell_gen_periodic_dummy_aura_target_on_target_AuraScript);

            void PeriodicTick(AuraEffect const* /*aurEff*/)
            {
                 GetHitUnit()->CastSpell(GetHitUnit(), uint32(GetEffectValue()), true);
            }

            void Register() override
            {
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_target_on_target_AuraScript::PeriodicTick, EFFECT_0, SPELL_AURA_PERIODIC_DUMMY);
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_target_on_target_AuraScript::PeriodicTick, EFFECT_1, SPELL_AURA_PERIODIC_DUMMY);
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_target_on_target_AuraScript::PeriodicTick, EFFECT_2, SPELL_AURA_PERIODIC_DUMMY);
            }
        };

        AuraScript* GetAuraScript() const override
        {
           return new spell_gen_periodic_dummy_aura_target_on_target_AuraScript();
        }
};

class spell_gen_periodic_dummy_aura_target_on_caster : public SpellScriptLoader
{
    public:
        spell_gen_periodic_dummy_aura_target_on_caster() : SpellScriptLoader("spell_gen_periodic_dummy_aura_target_on_caster") { }

        class spell_gen_periodic_dummy_aura_target_on_caster_AuraScript : public AuraScript
        {
            PrepareAuraScript(spell_gen_periodic_dummy_aura_target_on_caster_AuraScript);

            void PeriodicTick(AuraEffect const* /*aurEff*/)
            {
                GetHitUnit()->CastSpell(GetCaster(), uint32(GetEffectValue()));
            }

            void Register() override
            {
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_target_on_caster_AuraScript::PeriodicTick, EFFECT_0, SPELL_AURA_PERIODIC_DUMMY);
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_target_on_caster_AuraScript::PeriodicTick, EFFECT_1, SPELL_AURA_PERIODIC_DUMMY);
                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_periodic_dummy_aura_target_on_caster_AuraScript::PeriodicTick, EFFECT_2, SPELL_AURA_PERIODIC_DUMMY);
            }
        };

        AuraScript* GetAuraScript() const override
        {
           return new spell_gen_periodic_dummy_aura_target_on_caster_AuraScript();
        }
};
```

``` CPP
    new spell_gen_periodic_dummy_aura_caster_on_target();
    new spell_gen_periodic_dummy_aura_target_on_target();
    new spell_gen_periodic_dummy_aura_target_on_caster();
```

Aokromes commented:
@joschiwald this can be merged?
